### PR TITLE
Update snapshot cache and validate manually

### PR DESCRIPTION
**Pull request by kyessenov**
Working manual test of Envoy consuming ADS

Instructions to run:
```bash
term1> go run pkg/test/main/main.go
term2> envoy -c pkg/test/main/server_ads.yaml  --service-cluster "service-cluster" --service-node "service-node"
term3> while sleep 1; do curl localhost:9000/test; done
```

Signed-off-by: Kuat Yessenov <kuat@google.com>

### CODE REVIEW

File: pkg/cache/simple.go [hunk 1]
```diff
@@ -46,119 +49,169 @@ type SimpleCache struct {
 	mu sync.Mutex
 }
 
-// MakeSimpleCache initializes a simple cache.
+// Snapshot is an internally consistent snapshot of xDS resources.
+// Snapshots should have distinct versions per node group.
```
Comment (rshriram) [inline pkg/cache/simple.go]: Instead of saying per node group, why not just say per service cluster?

File: pkg/test/main/server_ads.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        address: 127.0.0.1
+        port_value: 18000
+    lb_policy: ROUND_ROBIN
+    http2_protocol_options: {}
+admin:
+  access_log_path: /dev/null
+  address:
+    socket_address:
+      address: 127.0.0.1
+      port_value: 19000
```
Comment (rshriram) [inline pkg/test/main/server_ads.yaml]: Where is the bit to start Envoy? If it’s manual can we just script those steps and dump script here? Helps others reproduce the test.

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -57,6 +59,14 @@ type Watch struct {
 	// close the corresponding stream.
 	Value chan Response
 
+	// Type is the response type.
+	Type ResponseType
+
+	// Names for monitored resources. LDS/CDS expect empty list since they are
+	// discovered globally. RDS/EDS expect explicitly enumerated resource names
+	// as dictated by LDS/CDS.
```
Comment (rshriram) [inline pkg/cache/cache.go]: Don’t understand. More explanation..

File: pkg/cache/cache.go [hunk 4]
```diff
@@ -82,23 +92,12 @@ type Response struct {
 	Canary bool
```
Comment (rshriram) [inline pkg/cache/cache.go]: I think we can get rid of this until we have that feature implemented

File: pkg/cache/simple.go [hunk 5]
```diff
@@ -46,119 +49,169 @@ type SimpleCache struct {
...
-		cache.versions[group] = make(map[ResponseType]int)
+	cache.snapshots[group] = snapshot
+
+	// trigger existing watches
+	if _, ok := cache.watches[group]; ok {
+		for _, watch := range cache.watches[group] {
+			respond(watch, snapshot, group)
+		}
+		// discard all watches; the client must request a new watch to receive updates and ACK/NACK
+		delete(cache.watches, group)
```
Comment (rshriram) [inline pkg/cache/simple.go]: Why trigger and then delete ?

File: pkg/cache/cache.go (hunk 3 above)
Comment (kyessenov) [reply pkg/cache/cache.go]: See the discussion on slack...

File: pkg/cache/cache.go (hunk 4 above)
Comment (kyessenov) [reply pkg/cache/cache.go]: This was explicitly requested by @junr03, so making provision to implement it.

File: pkg/cache/simple.go (hunk 1 above)
Comment (kyessenov) [reply pkg/cache/simple.go]: Node is not just a service cluster. There are labels and metadata inside. We may want to group by a combination of a cluster and envoy build for example.

File: pkg/cache/simple.go (hunk 5 above)
Comment (kyessenov) [reply pkg/cache/simple.go]: A trigger should make envoy request a new watch, no need to wait for cancellation.

File: pkg/test/main/server_ads.yaml (hunk 2 above)
Comment (kyessenov) [reply pkg/test/main/server_ads.yaml]: see PR description.

File: pkg/cache/cache.go (hunk 3 above)
Comment (rshriram) [reply pkg/cache/cache.go]: Could you summarize it and dump it here ?

Comment (kyessenov) [reply pkg/cache/cache.go]: It's pretty much what comment says. LDS/CDS request all resources. EDS/RDS request resources by name. We only reply to requested names since the rest gets ignored.

Comment (rshriram) [reply pkg/cache/cache.go]: And put this explanation above (which is clearer) in the comment :)

File: pkg/cache/cache.go (hunk 4 above)
Comment (rshriram) [reply pkg/cache/cache.go]: They are not using the discovery canary field. They are using EDS canary field, which is already covered by the fact that we are using the eds proto. cc @junr03

Comment (junr03) [reply pkg/cache/cache.go]: Sorry @kyessenov I don't remember talking to you about this? We don't currently use this field in the response.

Comment (kyessenov) [reply pkg/cache/cache.go]: Awesome, then I'll drop it for now. I remember on slack that RDS + canary is something that we want to support.

File: pkg/cache/cache.go (hunk 3 above)
Comment (kyessenov) [reply pkg/cache/cache.go]: Done

File: pkg/cache/resource.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	case *api.ClusterLoadAssignment:
+		return v.GetClusterName()
+	case *api.Cluster:
+		return v.GetName()
+	case *api.RouteConfiguration:
+		return v.GetName()
+	case *api.Listener:
+		return v.GetName()
+	default:
+		return ""
```
Comment (bndw) [inline pkg/cache/resource.go]: Should this return "unknown" for consistency with `ResponseType.String()`?

File: pkg/cache/simple.go [hunk 7]
```diff
@@ -46,119 +49,169 @@ type SimpleCache struct {
...
 	defer cache.mu.Unlock()
 
 	// update the existing entry
-	if _, ok := cache.responses[group]; !ok {
-		cache.responses[group] = make(map[ResponseType][]proto.Message)
-		cache.versions[group] = make(map[ResponseType]int)
+	cache.snapshots[group] = snapshot
+
+	// trigger existing watches
+	if _, ok := cache.watches[group]; ok {
```
Comment (bndw) [inline pkg/cache/simple.go]:
```go
if watches, ok := cache.watches[group]; ok {
    for _, watch := range watches {
    }
}
```

File: pkg/cache/simple.go [hunk 8]
```diff
@@ -46,119 +49,169 @@ type SimpleCache struct {
...
-	cache.responses[group][typ] = resources
+	return nil
+}
 
-	// increment version
-	cache.versions[group][typ]++
+// Respond to a watch with the snapshot value. The value channel should have capacity not to block.
+// TODO(kuat) this responds immediately and can cause the remote node to spin if it consistently fails to ACK the update
+func respond(watch Watch, snapshot Snapshot, group Key) {
+	typ := watch.Type
```
Comment (bndw) [inline pkg/cache/simple.go]: nit: I'm a fan of grouping declaration/assignments, but not sure if this is recommended. cc: @davecheney 

```go
var (
    typ        = watch.Type
    resources  = snapshot.resources[typ]
    version    = snapshot.version
)
```

**Comment by bndw** [comment]
Looks like I was too slow 😛

File: pkg/cache/simple.go (hunk 7 above)
Comment (kyessenov) [reply pkg/cache/simple.go]: thanks will follow up with this.

File: pkg/cache/resource.go (hunk 6 above)
Comment (kyessenov) [reply pkg/cache/resource.go]: an empty name is an invalid name (most likely) but a name `unknown` could be a legitimate resource name.

File: pkg/cache/simple.go (hunk 8 above)
Comment (kyessenov) [reply pkg/cache/simple.go]: let me know what the stylistic recommendation is :)
